registry/cache/rcache.go

```diff
@@ -328,9 +328,20 @@ func (c *cache) watch(w registry.Watcher) error {
 	defer w.Stop()
 
 	// manage this loop
+	quitLoop := make(chan interface{})
+	defer func() {
```

Comment:
can you comment this code, i don't understand it

Reply:
ok, actually it is because that even the watcher stops, [this goroutine](https://github.com/micro/go-micro/blob/master/registry/cache/rcache.go#L331) wont stop ,and we will make a new watcher and watch again, which leads to the number of goroutines increasing

Reply:
@vtolstov this function is a `defer` function so it's called when `watch()` returns. It unblocks (exits it) the following goroutine:

```Go
// manage this loop
	go func() {
		// wait for exit
		<-c.exit
		w.Stop()
	}()
```

which would otherwise be only unblocked only if `close(c.exit)` is called somewhere. Unless `close(c.exit)` is called somewhere this goroutine might really be "leaking".

My issue here is though. if we DO unblock the goroutine shown above, we will have **double** `w.Stop()` call, which I'm not sure can be done. What I mean by double:
* `watch` returns
* it calls `w.Stop()` (the first line in the `watch` function has the `defer` call)
* the "leak" goroutine gets unblocked and it **also** calls `w.Stop()`

Just my thoughts here...

Reply:
great, i think the w.Stop should be re-entrantable。

or

```
case <-quitLoop:
    return // wont call w.Stop() again
```